setup.py

```diff
@@ -69,8 +70,20 @@ def package_files(directory):
     return paths
 
 
-# this sets the __version__ variable
-exec(read_file(path.join("src", "cocotb", "_version.py")))
+__version__ = "2.0.0.dev0"
+if "dev" in __version__:
+    try:
+        rev = subprocess.check_output(
+            ["git", "rev-parse", "--short", "HEAD"], universal_newlines=True
+        ).strip()
+        __version__ += f"+{rev}"
+    except Exception as e:
+        print(e, file=sys.stderr)
+with open(path.join("src", "cocotb", "_version.py"), "w") as f:
+    f.write("# Package version\n")
+    f.write("# Generated by setup.py -- do not modify directly\n\n")
+    f.write(f'__version__ = "{__version__}"')
```

Comment:
sdist installation doesn't work, as it is installed from the `tar.gz` file, not a git repo.
You get this in the output:
```
  fatal: not a git repository (or any of the parent directories): .git
  Command '['git', 'rev-parse', '--short', 'HEAD']' returned non-zero exit status 128.
```

But, `_version.py` is included in the sdist, so use it in that case:

```suggestion
version_file_path = path.join("src", "cocotb", "_version.py")
__version__ = "2.0.0.dev0"
if "dev" in __version__:
    try:
        rev = subprocess.check_output(
            ["git", "rev-parse", "--short", "HEAD"], universal_newlines=True
        ).strip()
        __version__ += f"+{rev}"
    except Exception as e:
        # if this is not a git repository and _version.py already exists,
        # we are probably installing from an sdist, so use the existing _version.py
        if path.exists(version_file_path):
            exec(read_file(version_file_path))
        else:
            print(e, file=sys.stderr)
with open(version_file_path, "w") as f:
    f.write("# Package version\n")
    f.write("# Generated by setup.py -- do not modify directly\n\n")
    f.write(f'__version__ = "{__version__}"'
```

Reply:
How can I reproduce this?  I would have assumed catching the exception in the case of not installing from Git would have let setup.py finish and create _version.py just without the `+hash`>

Reply:
I edited my comment as it may have been unclear. It does install, but it is just `2.0.0.dev0` without the `+hash` suffix.
But the correct local version exists in the sdist `_version.py`, it is just overwritten by the default when it doesn't need to be.